llo/json_report_codec_test.go

```diff
@@ -16,6 +25,257 @@ import (
+		[]byte(fmt.Sprintf(packedJsonSigTemplate, `"abc$def#ghi!"`)),
+	}
+	for _, badSig := range badSigs {
+		f.Add(badSig)
+	}
+
+	// Set up the random number generator
+	rand.Seed(time.Now().UnixNano())
+
+	var codec JSONReportCodec
+	f.Fuzz(func(t *testing.T, data []byte) {
+		// test that it doesn't panic, don't care about errors
+		codec.Decode(data)
+		codec.Unpack(data)
+		codec.UnpackDecode(data)
+	})
+}
+
+func Test_JSONCodec_Properties(t *testing.T) {
+	properties := gopter.NewProperties(nil)
+
+	ctx := tests.Context(t)
+	cd := llotypes.ChannelDefinition{}
+	codec := JSONReportCodec{}
+
+	properties.Property("Encode/Decode", prop.ForAll(
+		func(r Report) bool {
+			b, err := codec.Encode(ctx, r, cd)
+			require.NoError(t, err)
+			r2, err := codec.Decode(b)
+			require.NoError(t, err)
+			return equalReports(r, r2)
+
+		},
+		gen.StrictStruct(reflect.TypeOf(&Report{}), map[string]gopter.Gen{
+			"ConfigDigest":                genConfigDigest(),
+			"SeqNr":                       genSeqNr(),
+			"ChannelID":                   gen.UInt32(),
+			"ValidAfterSeconds":           gen.UInt32(),
+			"ObservationTimestampSeconds": gen.UInt32(),
+			"Values":                      genStreamValues(),
+			"Specimen":                    gen.Bool(),
+		}),
+	))
+
+	properties.Property("Pack/Unpack", prop.ForAll(
+		func(digest types.ConfigDigest, seqNr uint64, report ocr2types.Report, sigs []types.AttributedOnchainSignature) bool {
+			b, err := codec.Pack(digest, seqNr, report, sigs)
+			require.NoError(t, err)
+			digest2, seqNr2, report2, sigs2, err := codec.Unpack(b)
+			require.NoError(t, err)
+
+			if digest != digest2 {
+				return false
+			}
+			if seqNr != seqNr2 {
+				return false
+			}
+			if !bytes.Equal(report, report2) {
+				return false
+			}
+			if len(sigs) != len(sigs2) {
+				return false
+			}
+			for i := range sigs {
+				if sigs[i].Signer != sigs2[i].Signer || !bytes.Equal(sigs[i].Signature, sigs2[i].Signature) {
+					return false
+				}
+			}
+			return true
+		},
+		genConfigDigest(),
+		genSeqNr(),
+		genSerializedReport(),
+		genSigs(),
+	))
+
+	properties.TestingRun(t)
+}
+
+func equalReports(r, r2 Report) bool {
+	if r.ConfigDigest != r2.ConfigDigest {
+		return false
+	}
+	if r.SeqNr != r2.SeqNr {
+		return false
+	}
+	if r.ChannelID != r2.ChannelID {
+		return false
+	}
+	if r.ValidAfterSeconds != r2.ValidAfterSeconds {
+		return false
+	}
+	if r.ObservationTimestampSeconds != r2.ObservationTimestampSeconds {
+		return false
+	}
+	if len(r.Values) != len(r2.Values) {
+		return false
+	}
+	for i := range r.Values {
+		if !equalStreamValues(r.Values[i], r2.Values[i]) {
+			return false
+		}
+	}
+	if r.Specimen != r2.Specimen {
+		return false
+	}
+	return true
+}
+
+func equalStreamValues(sv, sv2 StreamValue) bool {
+	if sv.Type() != sv2.Type() {
+		return false
+	}
+	m1, err := sv.MarshalBinary()
+	if err != nil {
+		panic(err)
+	}
+	m2, err := sv2.MarshalBinary()
+	if err != nil {
+		panic(err)
```

Comment:
How likely are these panics? For test files, it's usually pretty easy to plumb `*testing.T` to call `Fatal*` to be less disruptive:
```suggestion
func equalStreamValues(t *testing.T, sv, sv2 StreamValue) bool {
	if sv.Type() != sv2.Type() {
		return false
	}
	m1, err := sv.MarshalBinary()
	if err != nil {
		t.Fatal(err)
	}
	m2, err := sv2.MarshalBinary()
	if err != nil {
		t.Fatal(err)
```